### PR TITLE
TG-1158: Add a unit test for specialising Java generic types with array types

### PR DESCRIPTION
**Pull request by chrisr-diffblue**
This PR is built on top of https://github.com/diffblue/cbmc/pull/1543, and as such shows one commit (helper function) in common with PR1543

### CODE REVIEW

**Review by smowton — APPROVED** [review]

File: unit/java_bytecode/generate_concrete_generic_type/generate_java_generic_type.cpp [hunk 1]
```diff
@@ -160,3 +160,37 @@ SCENARIO(
...
+        "generic_field_array_instantiation",
+        "./java_bytecode/generate_concrete_generic_type");
+
+    REQUIRE(new_symbol_table.has_symbol(test_class));
+    const symbolt test_class_symbol=
+      new_symbol_table.lookup_ref(test_class);
+
+    REQUIRE(test_class_symbol.type.id()==ID_struct);
+    const typet test_field_type=
+      to_struct_type(test_class_symbol.type).components()[3].type();
```
Comment (thk123) [inline unit/java_bytecode/generate_concrete_generic_type/generate_java_generic_type.cpp]: Nit: use `require_type::require_component` to get the specific components type (not obvious that  `components()[3]` corresponds to `f`

Comment (thk123) [reply unit/java_bytecode/generate_concrete_generic_type/generate_java_generic_type.cpp]: In fact, is it actually pointing to `af` if the type is pointer to `array[reference]`?

File: unit/java_bytecode/generate_concrete_generic_type/generic_field_array_instantiation.java [hunk 2]
```diff
@@ -0,0 +1,11 @@
+public class generic_field_array_instantiation {
+
+    class generic<T> {
+        T gf;
+    }
+
+    generic<Float []> f;
```
Comment (thk123) [inline unit/java_bytecode/generate_concrete_generic_type/generic_field_array_instantiation.java:7]: The logic to specialise these is moving to just specialise inputs for functions - you may want to modify this test to directly call the specialisation logic for the generic field.

**Review by thk123 — CHANGES_REQUESTED** [review]
I'm not certain - but I think your test is looking at the array field rather than the generic field specialised with an array.

**Review by thk123 — APPROVED** [review]

**Pull request merged**